[Style] Convert clip-path and offset-path to use strong style types

weinig opened the pull request:
#### 2e33f4c127f8a672eecbc8a928c4801a0e75da3e
<pre>
[Style] Convert clip-path and offset-path to use strong style types
<a href="https://bugs.webkit.org/show_bug.cgi?id=294754">https://bugs.webkit.org/show_bug.cgi?id=294754</a>

Reviewed by Darin Adler.

Add strongly typed wrappers for the clip-path and offset-path
properties. These are both &quot;zero-cost&quot; wrappers around the
existing reference counted path operation types.

Converted the PathComputation and WindRuleComputation
interfaces to use the new CPO style invokers to match
the other invokers.

Added support for generic Variant support in TextStream and
made it so that elided tuples can be supported in TextStream
as well.

* Source/WTF/wtf/text/TextStream.h:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/KeyframeEffect.cpp:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/page/InteractionRegion.cpp:
* Source/WebCore/platform/animation/AcceleratedEffectValues.cpp:
* Source/WebCore/platform/graphics/FloatRoundedRect.cpp:
* Source/WebCore/platform/graphics/FloatRoundedRect.h:
* Source/WebCore/rendering/MotionPath.cpp:
* Source/WebCore/rendering/MotionPath.h:
* Source/WebCore/rendering/PathOperation.cpp:
* Source/WebCore/rendering/PathOperation.h:
* Source/WebCore/rendering/ReferencedSVGResources.cpp:
* Source/WebCore/rendering/ReferencedSVGResources.h:
* Source/WebCore/rendering/RenderBox.cpp:
* Source/WebCore/rendering/RenderElementInlines.h:
* Source/WebCore/rendering/RenderLayer.cpp:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
* Source/WebCore/rendering/RenderLayerModelObject.cpp:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
* Source/WebCore/rendering/svg/SVGRenderTreeAsText.cpp:
* Source/WebCore/rendering/svg/SVGRenderingContext.cpp:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceClipper.cpp:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceContainer.cpp:
* Source/WebCore/rendering/svg/legacy/SVGResources.cpp:
* Source/WebCore/rendering/svg/legacy/SVGResourcesCache.cpp:
* Source/WebCore/style/StyleAdjuster.cpp:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorCustom.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
* Source/WebCore/style/values/StyleValueTypes.h:
* Source/WebCore/style/values/masking/StyleClipPath.cpp: Added.
* Source/WebCore/style/values/masking/StyleClipPath.h: Added.
* Source/WebCore/style/values/motion/StyleOffsetPath.cpp: Added.
* Source/WebCore/style/values/motion/StyleOffsetPath.h: Added.
* Source/WebCore/style/values/shapes/StylePathComputation.h:
* Source/WebCore/style/values/shapes/StylePathFunction.cpp:
* Source/WebCore/style/values/shapes/StylePathFunction.h:
* Source/WebCore/style/values/shapes/StylePathOperationWrappers.cpp: Added.
* Source/WebCore/style/values/shapes/StylePathOperationWrappers.h: Added.
* Source/WebCore/style/values/shapes/StyleWindRuleComputation.h:
* Source/WebCore/svg/SVGClipPathElement.cpp:

Canonical link: <a href="https://commits.webkit.org/296560@main">https://commits.webkit.org/296560@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1ad15b4ef601dad6edb59d97a2c5eee766939d1f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108915 "48 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/28575 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/19000 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/114123 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/59260 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/110878 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/29259 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/37142 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/82763 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111863 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/23259 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/98103 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/63202 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22680 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/16241 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58821 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/101452 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/92634 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/16286 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/117243 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/107474 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35965 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/26577 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/91775 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/36337 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94367 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91582 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23314 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36489 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/14246 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/31834 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35864 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/41390 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/131749 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35565 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/35726 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38907 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/37250 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/page/InteractionRegion.cpp:
```diff
@@ -550,11 +550,10 @@ std::optional<InteractionRegion> interactionRegionForRenderedRegion(RenderObject
+    if (!hasRotationOrShear && WTF::holdsAlternative<Style::BasicShapePath>(style.clipPath()) && originalElement) {
         auto size = boundingSize(regionRenderer, transform);
-        auto path = styleClipPath->getPath(TransformOperationData(FloatRect(FloatPoint(), size)));
+        auto path = Style::tryPath(style.clipPath(), TransformOperationData(FloatRect(FloatPoint(), size)));
```
Given that this is inside code that knows it holds a `BasicShapePath`, I was wondering if there is some optimization or stronger type operation that can be done. Maybe not.

weinig replied:
Add try..() functions to make this work.

darinadler commented on Source/WebCore/platform/graphics/FloatRoundedRect.h:
```diff
@@ -154,6 +156,8 @@ class FloatRoundedRect {
 
     bool intersectionIsRectangular(const FloatRect&) const;
 
+    Path path() const;
```
I’m not sure I like the pattern here.

Does `FloatRect` have a `path` function in it? What about `IntRect`? If we want to convert something that can be added to a path into a path with just that single geometric figure in it, wouldn’t we want to do that with a template function rather than adding a member function to each type?

I didn’t necessarily spot all the new code that depends on this, so I’m not sure how much this makes the code more elegant or if it makes a template abstraction possible. At least some of the call sites I found don’t really seem like they need to be the way they are.

weinig replied:
I think having `FloatRect` and `IntRect` having `path()` functions sounds good actually. 

The alternative, to make the call sites a bit more weirdly is to give Path() a constructor that takes a FloatRoundedRect (and ones for FloatRect and IntRect), which would also be ok with me. 

Its all just to avoid the multiple lines needed to construct and add the one shape you want.

darinadler commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -1571,33 +1568,30 @@ bool RenderBox::hitTestClipPath(const HitTestLocation& hitTestLocation, const La
+            auto referenceBoxRect = this->referenceBoxRect(clipPath.referenceBox());
+            auto path = Style::path(clipPath.shape(), referenceBoxRect);
+            auto windRule = Style::windRule(clipPath.shape());
+            return path.contains(hitTestLocationInLocalCoordinates, windRule);
```
```suggestion
            auto& shape = clipPath.shape();
            auto path = Style::path(shape, referenceBoxRect(clipPath.referenceBox()));
            return path.contains(hitTestLocationInLocalCoordinates, Style::windRule(shape));
```

darinadler commented on Source/WebCore/rendering/RenderBox.cpp:
```diff
@@ -1571,33 +1568,30 @@ bool RenderBox::hitTestClipPath(const HitTestLocation& hitTestLocation, const La
+                return true;
+            if (!hitsClipContent(*element))
+                return false;
+            return true;
```
```suggestion
            return !is<SVGClipPathElement>(element) || !element->renderer() || hitsClipContent(*element);
```

darinadler commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -3476,24 +3477,29 @@ bool RenderLayer::setupFontSubpixelQuantization(GraphicsContext& context, bool&
+                auto shapeRect = computeRoundedRectForBoxShape(clipPath.referenceBox(), *box).pixelSnappedRoundedRectForPainting(renderer().document().deviceScaleFactor());
+                shapeRect.move(offsetFromRoot);
+                Path path;
+                path.addRoundedRect(shapeRect);
```
This looks like it might be a place you were planning to use `FloatRoundedRect::path`

darinadler commented on Source/WebCore/rendering/RenderLayerModelObject.cpp:
```diff
@@ -627,24 +632,25 @@ bool RenderLayerModelObject::pointInSVGClippingArea(const FloatPoint& point) con
+            auto referenceBox = clipPathReferenceBox(clipPath.referenceBox());
+            if (!referenceBox.contains(point))
+                return false;
+            return Style::path(clipPath.shape(), referenceBox).contains(point, Style::windRule(clipPath.shape()));
```
Can this be done more optimally for some types of basic shape paths?

weinig replied:
For sure.

darinadler commented on Source/WebCore/rendering/RenderLayerModelObject.cpp:
```diff
@@ -627,24 +632,25 @@ bool RenderLayerModelObject::pointInSVGClippingArea(const FloatPoint& point) con
+            auto referenceBox = clipPathReferenceBox(clipPath.referenceBox());
+            if (!referenceBox.contains(point))
+                return false;
+            return FloatRoundedRect { referenceBox }.path().contains(point);
```
What’s the point of making a rounded rectangle if the radii are always zero? Won’t this always return true?

weinig replied:
I'm not entirely sure, but its just a direct translation of the old code:

```c++
        auto referenceBox = clipPathReferenceBox(clipPath->referenceBox());
        if (!referenceBox.contains(point))
            return false;
        return clipPath->pathForReferenceRect(referenceBox).contains(point, clipPath->windRule());
```

where what `clipPath->pathForReferenceRect` does is:

```c++
    Path pathForReferenceRect(const FloatRoundedRect& boundingRect) const
    {
        Path path;
        path.addRoundedRect(boundingRect);
        return path;
    }
```

darinadler commented on Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
```diff
@@ -410,51 +410,52 @@ inline FloatRect clipPathReferenceBox(const RenderElement& renderer, CSSBoxType
+            auto referenceBox = clipPathReferenceBox(renderer, clipPath.referenceBox());
+            if (!referenceBox.contains(point))
+                return false;
+            return FloatRoundedRect { referenceBox }.path().contains(point);
```
What’s the point of making a rounded rectangle if the radii are always zero? Won’t this always return true?

weinig replied:
Just maintaining the existing code's structure. 

Old code was:

```c++
clipPath->pathForReferenceRect(FloatRoundedRect { referenceBox }).contains(point)
```

darinadler commented on Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
```diff
@@ -410,51 +410,52 @@ inline FloatRect clipPathReferenceBox(const RenderElement& renderer, CSSBoxType
+        },
+        [&](const Style::BoxPath& clipPath) {
+            auto referenceBox = clipPathReferenceBox(renderer, clipPath.referenceBox());
+            context.clipPath(FloatRoundedRect { referenceBox }.path());
```
Why convert this to a rounded rectangle if the radii are always zero? Can’t we convert a `FloatRect` to a `Path` without involving `FloatRoundedRect`?

weinig replied:
Again, trying not to mix potentially behavior changes with factoring.

darinadler commented on Source/WebCore/style/values/motion/StyleOffsetPath.h:
```diff
@@ -0,0 +1,133 @@
+    bool isBasicShape() const { return is<ShapePathOperation>(operation); }
+    bool isBox() const { return is<BoxPathOperation>(operation); }
+
+    template<typename T> bool holdsAlternative() const;
```
```suggestion
    template<typename> bool holdsAlternative() const;
```

darinadler commented on Source/WebCore/style/values/masking/StyleClipPath.h:
```diff
@@ -0,0 +1,127 @@
+    bool isBasicShape() const { return is<ShapePathOperation>(operation); }
+    bool isBox() const { return is<BoxPathOperation>(operation); }
+
+    template<typename T> bool holdsAlternative() const;
```
```suggestion
    template<typename> bool holdsAlternative() const;
```

darinadler commented on Source/WebCore/style/values/shapes/StylePathComputation.h:
```diff
@@ -39,16 +39,19 @@ namespace Style {
+        return PathComputation<StyleType>{}(value, reference);
+    }
+};
+inline constexpr PathComputationInvoker path{};
```
```suggestion
inline constexpr PathComputationInvoker path { };
```

weinig committed:
[Style] Convert clip-path and offset-path to use strong style types
https://bugs.webkit.org/show_bug.cgi?id=294754

Reviewed by Darin Adler.

Add strongly typed wrappers for the clip-path and offset-path
properties. These are both "zero-cost" wrappers around the
existing reference counted path operation types.

Converted the PathComputation and WindRuleComputation
interfaces to use the new CPO style invokers to match
the other invokers.

Added support for generic Variant support in TextStream and
made it so that elided tuples can be supported in TextStream
as well.

* Source/WTF/wtf/text/TextStream.h:
* Source/WebCore/CMakeLists.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/KeyframeEffect.cpp:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/page/InteractionRegion.cpp:
* Source/WebCore/platform/animation/AcceleratedEffectValues.cpp:
* Source/WebCore/platform/graphics/FloatRoundedRect.cpp:
* Source/WebCore/platform/graphics/FloatRoundedRect.h:
* Source/WebCore/rendering/MotionPath.cpp:
* Source/WebCore/rendering/MotionPath.h:
* Source/WebCore/rendering/PathOperation.cpp:
* Source/WebCore/rendering/PathOperation.h:
* Source/WebCore/rendering/ReferencedSVGResources.cpp:
* Source/WebCore/rendering/ReferencedSVGResources.h:
* Source/WebCore/rendering/RenderBox.cpp:
* Source/WebCore/rendering/RenderElementInlines.h:
* Source/WebCore/rendering/RenderLayer.cpp:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
* Source/WebCore/rendering/RenderLayerModelObject.cpp:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
* Source/WebCore/rendering/svg/SVGRenderTreeAsText.cpp:
* Source/WebCore/rendering/svg/SVGRenderingContext.cpp:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceClipper.cpp:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceContainer.cpp:
* Source/WebCore/rendering/svg/legacy/SVGResources.cpp:
* Source/WebCore/rendering/svg/legacy/SVGResourcesCache.cpp:
* Source/WebCore/style/StyleAdjuster.cpp:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleExtractorConverter.h:
* Source/WebCore/style/StyleExtractorCustom.h:
* Source/WebCore/style/StyleExtractorSerializer.h:
* Source/WebCore/style/values/StyleValueTypes.h:
* Source/WebCore/style/values/masking/StyleClipPath.cpp: Added.
* Source/WebCore/style/values/masking/StyleClipPath.h: Added.
* Source/WebCore/style/values/motion/StyleOffsetPath.cpp: Added.
* Source/WebCore/style/values/motion/StyleOffsetPath.h: Added.
* Source/WebCore/style/values/shapes/StylePathComputation.h:
* Source/WebCore/style/values/shapes/StylePathFunction.cpp:
* Source/WebCore/style/values/shapes/StylePathFunction.h:
* Source/WebCore/style/values/shapes/StylePathOperationWrappers.cpp: Added.
* Source/WebCore/style/values/shapes/StylePathOperationWrappers.h: Added.
* Source/WebCore/style/values/shapes/StyleWindRuleComputation.h:
* Source/WebCore/svg/SVGClipPathElement.cpp:

Canonical link: https://commits.webkit.org/296560@main
